mantidproject/mantid
# Extend attenuation correction in AnvredCorrection higher muR

**Description of work.**
Have extended the attenuation correction to higher muR by fitting function ln(A) to a polynmial as a function of muR at fixed theta (as done in [1]) - but using the tabulated A* values (A* = 1/A) from [2] which goes to higher muR.

This appears to have also fixed a bug - as the agreement with the Monte-Carlo is much better now...

![image](https://user-images.githubusercontent.com/55979119/140786164-90aca104-adba-4738-88b0-a4cb056d3dec.png)


The correction factor to apply is now found by interpolation of A  (rather than A*) as recommended in [1]. In addition I have replaced the linear interpolation with interpolation using the Eq.
`A(th) = L0 + L1*(sin(th)^2)`
This captures better the smooth variation of the `A(th)` and exhibits roughly the right functional form (especially at low muR). Indeed in [1] they suggest that it is possible to fit ` A(th) = sum_{i=0}^{6} L_i *(sin(th)^(2i))`.
I've confirmed that this does indeed work better (see green markers in residual plot), below I interpolate using fit parameters from `theta +/- 5 deg` from the black data points at three theta values.
 
![image](https://user-images.githubusercontent.com/55979119/140897446-4db7b6cf-0925-446e-a177-9145bcaa1815.png)


**To test:**

(1) Check that the values of Astar agree with the tabulated values in [2]. Choose some theta and muR values to check from table and put them in the variables of this script. Try muR > 2.5 (i.e. greater than previous limit). 

```
theta = 50
muR = 1
CreateSampleWorkspace(OutputWorkspace='ws', Function='Powder Diffraction', 
    XMin=2000, XMax=20000, BinWidth=200, NumBanks=1, BankPixelWidth=1, PixelSpacing=0.02,
    WorkspaceType="Histogram")
EditInstrumentGeometry(Workspace='ws', L2=0.5, Polar=2*theta)
ConvertUnits(InputWorkspace='ws', OutputWorkspace='ws', Target='Wavelength')
# 1 cm sphere
shape = '''<sphere id="V-sphere">
    <centre x="0.0"  y="0.0" z="0.0" />
    <radius val="0.01"/>
    </sphere>'''
SetSample('ws', Geometry={'Shape': 'CSG', 'Value': shape})
SetSampleMaterial('ws', SampleNumberDensity=1.0,
     ScatteringXSection = muR, AttenuationXSection=0.0, ChemicalFormula='V') 
wsphere = SphericalAbsorption(InputWorkspace='ws', OutputWorkspace='attenuation_SphericalAbsorption')
w = MonteCarloAbsorption(InputWorkspace='ws', OutputWorkspace='attenuation_MonteCarlo', EventsPerPoint=5000)

Astar_sphere = 1/wsphere.readY(0)[0]
Astar_mc = 1/w.readY(0)[0]
print(Astar_sphere, Astar_mc)  # compare this with Dwiggins
```

The values should be within <~ 0.5% of the Weber values [2] for most theta and muR (and indeed consistent within the errorbars of the monte-carlo).

There should be a warning if muR > 9 and the result won't agree as well with the Monte-Carlo particularly at large theta (however for neutrons samples rarely have muR>3).

(2) Test the interpolation by picking a theta that isn't a multiple of 5 - it should still be consistent with the Monte-Carlo 

(3) Try adding a finite absorption cross-section to compare the Monte-Carlo curve to the `SphericalAbsorption`. Note the latter is a histogram, so plot it first and turn off normalisation (right-click on axes -> Normalisation -> None) before overplotting the Monte-Carlo curve (or plot using this script).
```
SetSampleMaterial('ws', SampleNumberDensity=1.0,
     ScatteringXSection = 1.0, AttenuationXSection=1.0, ChemicalFormula='V') 
attenuation_SphericalAbsorption = SphericalAbsorption(InputWorkspace='ws', OutputWorkspace='attenuation_SphericalAbsorption')
attenuation_MonteCarlo = MonteCarloAbsorption(InputWorkspace='ws', OutputWorkspace='attenuation_MonteCarlo', EventsPerPoint=5000)

fig, axes = plt.subplots(edgecolor='#ffffff', num='attenuation_SphericalAbsorption-1', subplot_kw={'projection': 'mantid'})
axes.plot(attenuation_SphericalAbsorption, color='#1f77b4', distribution=True, label='attenuation_SphericalAbsorption: spec 1', marker='.', markersize=3.0, wkspIndex=0)
axes.errorbar(attenuation_MonteCarlo, capsize=2.0, color='#ff7f0e', elinewidth=1.0, label='attenuation_MonteCarlo: spec 1', marker='.', markersize=3.0, wkspIndex=0)
axes.set_title('attenuation_SphericalAbsorption')
axes.set_xlabel('Wavelength ($\\AA$)')
axes.set_ylabel('Attenuation factor')
legend = axes.legend(fontsize=8.0).set_draggable(True).legend
fig.show()
```

[1]  https://scripts.iucr.org/cgi-bin/paper?a11726
 Dwiggins, C. W. "Rapid calculation of X-ray absorption correction factors for spheres to an accuracy of 0.05%." Acta Crystallographica Section A: Crystal Physics, Diffraction, Theoretical and General Crystallography 31.3 (1975): 395-396.

[2] https://journals.iucr.org/b/issues/1969/06/00/a06780/a06780.pdf
Weber, K. "Eine neue Absorptionsfaktortafel fur kugelformige Proben." Acta Crystallographica Section B: Structural Crystallography and Crystal Chemistry 25.6 (1969): 1174-1178.

Fixes #31342 






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
